### ISSUE TITLE
First run of /simplesaml throw TypeError in SessionHandlerPHP

### ISSUE THREAD

**Issue by Etienne-Schmitt** [simplesamlphp/simplesamlphp#1736]
# **Help us to keep this issue-tracker clean!  For questions or support, please refer to our [mailing lists](https://simplesamlphp.org/lists)**

**Describe the bug**
`TypeError` is thrown when starting the `/simplesaml` page when run with php 8.2

**To Reproduce**
1. Config everything like explained in the installation documentation
2. Go to /simplesaml
3. See error

**Expected behavior**
This `TypeError` should not be thrown and page should load in php 8.2

**Screenshots or logs**
```
Fatal error: Uncaught TypeError: Cannot assign null to property SimpleSAML\SessionHandlerPHP::$cookie_name of type string in
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/SessionHandlerPHP.php:51 Stack trace: #0
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/SessionHandler.php(143): SimpleSAML\SessionHandlerPHP->__construct() #1 
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/SessionHandler.php(44): SimpleSAML\SessionHandler::createSessionHandler() #2 
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/Session.php(333): SimpleSAML\SessionHandler::getSessionHandler() #3 
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/Session.php(263): SimpleSAML\Session::getSession() #4 
/var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/Logger.php(323): SimpleSAML\Session::getSessionFromRequest() #5 [internal function]:
SimpleSAML\Logger::shutdown() #6 {main} thrown in /var/www/html/sn_develop/framework/vendor/simplesamlphp/simplesamlphp/src/SimpleSAML/SessionHandlerPHP.php on line 51
```

**Additional context**

`SessionHandlerPHP::$cookie_name` is typed as a non nullable `string` type: https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L26

But is compared to nulls values in two lines in the code:

1. Configuration::getOptionalString() can return `null`
https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L51
2. $this->cookie_name is compared with type checking `===` with null https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L57

Maybe SessionHandlerPHP::$cookie_name should be typed as nullable string `?string` ? in this case, the others usage of `SessionHandlerPHP::$cookie_name` should correctly typed checked before using it:

1. [`\session_name()` can return `false`](https://www.php.net/manual/en/function.session-name.php) which will lead to `TypeError` https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L76
2. Usage of an absent key in a array can 'throw' NOTICE in the logs:
https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L184 https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L188
3. `SessionHandlerPHP::getSessionCookieName()` only return non nullable `string` type
https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L207
4. [1st params of `array_key_exists()` should be `string|int` or `TypeError` will be throw](https://www.php.net/manual/en/function.array-key-exists)
https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L264

**Comment by Etienne-Schmitt** [comment]
I forgot to add this:

I may do a PR fixing this issue when I've got some free time...

Sorry to forgot it before hahaha

**Comment by steinmb** [comment]
This is on HEAD right?

**Comment by Etienne-Schmitt** [comment]
> This is on HEAD right?

Yes, HEAD/master branch, now than you talk about it, i've looked at the issues and PR but didn't thing to look at the others branch...

**Comment by tvdijen** [comment]
How is your `session.phpsession.cookiename` set?

**Comment by Etienne-Schmitt** [comment]
> How is your `session.phpsession.cookiename` set?
> 
> The problem is in this line;
> 
> https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L51
> 
> However, I don't see how this can evaluate to `null`... Even though the logics should probably say `??` instead of `?:`...
I've looked at the configuration I've been given and... `session.phpsession.cookiename` was set to `null`...

But I really thing than the class `SessionHandlerPHP` need some smalls tweakings, why compare to null:
https://github.com/simplesamlphp/simplesamlphp/blob/ac4ae28ee1270391a34f5271a3f8aeb44d2d311c/src/SimpleSAML/SessionHandlerPHP.php#L57
if it can never be null because of non nullable type `string` instead of `?string`, the same config used to works in 1.19.7 without this problems...

**Comment by steinmb** [comment]
Mmm, that line looks iffy. Condition is always 'false' because '$this->cookie_name' is evaluated at this point. Could be simplified to only

```php
if (session_name() === $this->cookie_name) {}
```

**Comment by thijskh** [comment]
Maybe taking it one level up. If a variable is set in config but `null`, getOptionalString will now consider it as "set" and not use the "default" from the second argument.

One might argue that we change getOptionalString to consider a key that is present in config but assigned `null` to not be set to a value.

**Comment by tvdijen** [comment]
I would change the line to:
```php
$this->cookie_name = ($config->getOptionalString('session.phpsession.cookiename') ?? ini_get('session.name')) ?? 'PHPSESSID';
```

**Comment by thijskh** [comment]
So, solving it only for this specific case and not the general case?

**Comment by tvdijen** [comment]
If you're changing the behaviour you have to go over all the cases where this method is used (and all the other getOptional* methods) to make sure you're not breaking anything.. It's a _lot_ of work and it may not be worth it.

**Comment by thijskh** [comment]
Fine to fix the immediate bug in your proposal.

But I think that having options where 'unset' means something different than 'null' is endlessly confusing for both users and people working on the code so we should 'rationalize' the config by removing this distinction if it still exists anywhere. I don't think the distinction currently exists in the examples you provide (even though NameIdPolicy is a mess currently, unset still seems the same als null).

**Comment by steinmb** [comment]
Cannot wait to use PHP 8.2 where null, false, and true are stand-alone types :)
